File: FateGrandAutomata/Gestures/RootGestures.cs
using System;
using CoreAutomata;

namespace FateGrandAutomata
{
    public class RootGestures : IGestureService
    {
        readonly SuperUser _superUser;

        public RootGestures(SuperUser SuperUser)
        {
            _superUser = SuperUser;
        }

        const string InputCommand = "/system/bin/input";

        public void Swipe(Location Start, Location End)
        {
            _superUser.SendCommand($"{InputCommand} swipe {Start.X} {Start.Y} {End.X} {End.Y} {GestureTimings.SwipeDurationMs}");
        }

        public void Click(Location Location)
        {
            ContinueClick(Location, 1);
        }

        public void ContinueClick(Location Location, int Times)
        {
            Times = Math.Max(1, Times / 5);
            const int clickDuration = 1;

            while (Times-- > 0)
            {
                //_superUser.SendCommand($"{InputCommand} tap {Location.X} {Location.Y}");
                _superUser.SendCommand($"{InputCommand} swipe {Location.X} {Location.Y} {Location.X} {Location.Y} {clickDuration}");
            }
        }

        public void Dispose() { }
    }
}

File: app/src/main/java/com/mathewsachin/fategrandautomata/root/RootGestures.kt
package com.mathewsachin.fategrandautomata.root

import com.mathewsachin.libautomata.IGestureService
import com.mathewsachin.libautomata.Location
import com.mathewsachin.fategrandautomata.util.swipeDuration
import kotlin.math.max

private const val InputCommand = "/system/bin/input"

/**
 * Performs gestures using shell commands, which are only accessible as superusers (root).
 */
class RootGestures(private val SuperUser: SuperUser) : IGestureService {
    override fun swipe(Start: Location, End: Location) {
        SuperUser.sendCommand("$InputCommand swipe ${Start.X} ${Start.Y} ${End.X} ${End.Y} $swipeDuration")
    }

    override fun click(Location: Location, Times: Int) {
        val times = max(1, Times / 5)
        val clickDuration = 1

        repeat(times) {
            //_superUser.SendCommand($"{InputCommand} tap {Location.X} {Location.Y}");
            SuperUser.sendCommand("$InputCommand swipe ${Location.X} ${Location.Y} ${Location.X} ${Location.Y} $clickDuration")
        }
    }

    override fun close() {}
}